<a href="https://colab.research.google.com/github/chaitanysaraswat/Advanced-Analytics-Using-Statistics/blob/main/lecture10.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [20]:
import numpy as np
import pandas as pd
import warnings
warnings.filterwarnings("ignore")
pd.set_option('display.max_columns', None)
from numpy import random
from collections import Counter
from numpy.linalg import inv
from numpy.linalg import eig
import matplotlib
from matplotlib import pyplot as plt
import seaborn as sns
import pylab
from pylab import legend
from pylab import plot, show, title, xlabel, ylabel
import scipy
from scipy import stats
from scipy.stats import binom,poisson,norm,t,expon,f
from sklearn.model_selection import train_test_split
import statsmodels.api as sample_data
from statsmodels.formula.api import ols
import statsmodels.stats.multicomp
from statsmodels.stats.weightstats import ttost_paired
import statsmodels.stats.weightstats as ssw
from statsmodels.stats.multicomp import pairwise_tukeyhsd

Inferential:
- 1 sample ztest
- 2 sample ztest different sample  size is allowed
- sample ztest first diff than avg
- paired ttest first avg than diff b/w the two   

paired t test:
- this test is used when we have two paired samples avg diff b/w paired values
  is compared against a claim value to establish whether the diff statstically significant or not.
 - the keyword here is "before-after".
 - sample size needs to be size  source of data needs to be same

Time taken by individual will be coming by atleast by  2min

h0(null hypothesis) before -after >=2min

Q procesing will be reduced by atleast 1min. h0: before-after >=1min right

Q performance has reduced by  not more than 1min. h0: before-after<=1min left tail test

after-before>=10


In [2]:
before=np.array([12,14,10,9,8,13])
after=np.array([11.8,9.7,11.3,8.2,7.2,12.3])


In [3]:
d=(before-after)
np.mean(d)

np.float64(0.9166666666666665)

h0  avg(before-after)>2

now we wil check whether the 0.911 and >2 above is matter of chance variation or not

In [9]:

ssw.ttost_paired(before, after, low=-2, upp=2)

(np.float64(0.10453031697930062),
 (np.float64(3.8804663986357473),
  np.float64(0.005818213196929444),
  np.float64(5.0)),
 (np.float64(-1.4413160909218492),
  np.float64(0.10453031697930062),
  np.float64(5.0)))

Pvalue>0.05 so we do not reject null hypothesis

df is degree of freedom which is (no of values-1)

In [4]:
scipy.stats.ttest_1samp(before-after,popmean=2,alternative="less")

TtestResult(statistic=np.float64(-1.4413160909218492), pvalue=np.float64(0.10453031697930062), df=np.int64(5))

In [ ]:
(np.mean(d)-2)/((np.std(d,ddof=1))/6**0.5)
# (mean(diff)-claim_mean)/std(d)/number**0.5

# ANOVA ( ANalysis Of Variance )

pupose:
- this test is used when we have multiple samples.
- this test will compare all the pair of the means and check wheather they can be considered equal or not.
- standard null hpothesis all the pair means are equal

In [11]:
dfA = pd.DataFrame({'Col1':np.repeat('A',7),'Col2':np.random.randint(10,30,7)}) # col1 will have a is 7 no of times A and col 2 will 7 random value
dfB = pd.DataFrame({'Col1':np.repeat('B',9),'Col2':np.random.randint(20,40,9)}) # col1 will have a is 9 no of times A and col 2 will 9 random value
dfC = pd.DataFrame({'Col1':np.repeat('C',8),'Col2':np.random.randint(15,30,8)}) # col1 will have a is 8 no of times A and col 2 will 8 random value
df = pd.concat([dfA,dfB,dfC])


In [12]:
df

,Col1,Col2
0,A,24
1,A,16
2,A,11
3,A,16
4,A,27
5,A,25
6,A,11
0,B,33
1,B,23
2,B,20


h0 mean(a)=mean(b)

Anova below method
- Test Statstics will be f below method
- f=variations b/w samples/variations within sample
- deviation of each datapoint from its own sample mean is variation within sample.it is denoted by w in pic
-  variation each sample mean from the grand mean which is denoted by b in pic


- as the value of test statstics increases it indicates that we are going away from epected result so p value will decrease
case1: if between variation of samples is more and within variation is less in sample TTHEN VALUE OF F WILL INCREASE



# -** f distribution is always ratio of two variances**


In [18]:
mod1 = ols('Col2 ~ Col1',df).fit()
tbl = sm.stats.anova_lm(mod1)
print(tbl)

            df      sum_sq     mean_sq         F    PR(>F)
Col1       2.0  304.813492  152.406746  5.951047  0.008963
Residual  21.0  537.811508   25.610072       NaN       NaN


In [21]:
print(pairwise_tukeyhsd(df.Col2, df.Col1))

 Multiple Comparison of Means - Tukey HSD, FWER=0.05 
group1 group2 meandiff p-adj   lower    upper  reject
-----------------------------------------------------
     A      B    7.873 0.0148   1.4447 14.3013   True
     A      C   1.0536  0.915  -5.5481  7.6553  False
     B      C  -6.8194 0.0294 -13.0176 -0.6213   True
-----------------------------------------------------


ANOVA using excel sheet

-

In [57]:
df1=pd.read_excel('/content/drive/MyDrive/My pitch/data_anova (1).xlsx')

In [69]:
df1.head(20)

,Col1,Col2
0,A,10
1,A,23
2,A,14
3,A,25
4,A,22
5,A,11
6,A,12
7,B,26
8,B,39
9,B,29


In [72]:
df1.index=np.arange(df.shape[0])

source->db->etl->t->spark,pythoncollab,grafana,kibana,prometheus,powerbi,tableau

In [73]:
grp1=df1.groupby("Col1")

In [74]:
sm=grp1.agg('mean')
sm

,Col2
Col1,
A,16.714286
B,31.222222
C,23.250000


In [75]:
gm=np.mean(df1.Col2)
gm

np.float64(24.333333333333332)

In [76]:
sm.loc['A']

,A
Col2,16.714286


In [77]:
7*(sm.loc['A'] - gm)**2 + 9*(sm.loc['B'] - gm)**2 + 8*(sm.loc['C'] - gm)**2

,0
Col2,842.849206


In [86]:
sampA = df1.Col2[np.where(df.Col1=='A')[0]]
sampB = df1.Col2[np.where(df.Col1=='B')[0]]
sampC = df1.Col2[np.where(df.Col1=='C')[0]]

In [88]:
sum((np.array(sampA) - np.array(sm.loc['A']))**2) + sum((np.array(sampB) - np.array(sm.loc['B']))**2) + sum((np.array(sampC) - np.array(sm.loc['C']))**2)


np.float64(684.484126984127)

ANOVA TABLE DESCRIPTION:

COL1 SUM SQ COLUM=>842.849(B/W SAMPLES)

RESIDUAL SUMSQ COLUMN =>684 (WITHIN SAMPLES)

COL1

mean squared=sample stats/degree of freedom


In [91]:
import statsmodels.api as sm
from statsmodels.formula.api import ols
mod1=ols('Col2 ~ Col1',df).fit()
tbl=sm.stats.anova_lm(mod1)
tbl

,df,sum_sq,mean_sq,F,PR(>F)
Col1,2.0,304.813492,152.406746,5.951047,0.008963
Residual,21.0,537.811508,25.610072,NaN,NaN


one way anova is used when the response is continuos data
sampling will done on categorical data

Assumption one anova :

- data for all the sample need to be normally distributed
- variance of all the samples need to be equal

multiple anova:

whether gender seffects salary
whether

In [92]:
zoo = [34,43,57,40,85,68,67,53,41,24,42,52]
supp=[1,1,1,1,2,2,2,2,3,3,3,3]
lake=["R","D","R","D","R","D","R","D","R","D","R","D"]
df=pd.DataFrame({'Resp':zoo,"Factor1":supp,"Factor2":lake})